### PR TITLE
Add support for concurrency mocking

### PR DESCRIPTION
**Pull request by ailtonvivaz**
This PR implements support for parsing async methods and closures as discussed in #229

### CODE REVIEW

**Review by andrewchang-bird — COMMENTED** [review]
Thanks for making this PR—looks really good. It seems that Swift is choosing the incorrect overload because the constraint system heavily penalizes mismatched calling contexts ([coderef](https://github.com/apple/swift/blob/83484f85163bdfbe743f0a840451982268644d3c/lib/Sema/ConstraintSystem.cpp#L3008-L3016)). In the case of XCTests, we’re calling a synchronous method overload from an async context.

I’d like to avoid additional flavors of `given` and `verify` to keep the testing API surface as small as possible. After experimenting a bit, I think the cleanest API that also makes sense semantically is something like this:

```swift
given(await myMock.doSomethingAsync()).willReturn(someValue)
```

It does require marking the `Mockable<...>` framework declarations as async. Let me know what you think.

File: Sources/MockingbirdGenerator/Generator/Templates/MethodTemplate.swift [hunk 1]
```diff
@@ -336,13 +341,20 @@ class MethodTemplate: Template {
   }
   
   lazy var returnTypeAttributesForMocking: String = {
-    if method.attributes.contains(.rethrows) { return " rethrows" }
-    if method.attributes.contains(.throws) { return " throws" }
-    return ""
+    var attributes = ""
+    if method.attributes.contains(.async) {
+      attributes += " async"
+    }
```
[the remark this answers is not in the document]
Comment (ailtonvivaz) [reply Sources/MockingbirdGenerator/Generator/Templates/MethodTemplate.swift]: Fixed

File: Tests/MockingbirdTests/Framework/StubbingAsyncTests.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        line: UInt = #line,
+        _ errorHandler: (_ error: Error) -> Void = { _ in }
+    ) async {
+        do {
+            _ = try await expression()
+            XCTFail(message(), file: file, line: line)
+        } catch {
+            errorHandler(error)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ailtonvivaz) [reply Tests/MockingbirdTests/Framework/StubbingAsyncTests.swift]: Fixed

File: Sources/MockingbirdGenerator/Generator/Templates/MethodTemplate.swift [hunk 3]
```diff
@@ -336,13 +341,20 @@ class MethodTemplate: Template {
...
+    if method.attributes.contains(.rethrows) { attributes += " rethrows" }
+    if method.attributes.contains(.throws) { attributes += " throws" }
```
[the remark this answers is not in the document]
Comment (ailtonvivaz) [reply Sources/MockingbirdGenerator/Generator/Templates/MethodTemplate.swift]: Fixed

**Comment by ailtonvivaz** [comment]
Nice, you're right. That's a shame, unfortunately, because the method has the `@_disfavoredOverload` attribute. 😕 

Wow, your idea sounds great.
I am implementing and I will commit soon.
Thanks for the workaround.

There is just one more thing to considering. Do you think this "syntax" should be extended to throwable methods?
```swift
given(try await myMock.doSomethingAsyncAndThrowable()).willReturn(someValue)
```
Even for non async methods?
```swift
given(try myMock.doSomethingThrowable()).willReturn(someValue)
```

Or just for `await`keyword?

**Comment by ailtonvivaz** [comment]
Workaround implemented 🎉
Now, it's possible mock async methods using:
```swift
given(await myMock.doSomethingAsync()).willReturn(someValue)
```
And verify like before as well.
```swift
verify(await myMock.doSomethingAsync()).wasCalled()
```

**Review by andrewchang-bird — APPROVED** [review]
Nice work! The changes look good to me.

> Do you think this "syntax" should be extended to throwable methods?

If Swift’s overload handling changes we might be able to remove `await` as a requirement when stubbing/verifying, so let’s keep it simple and hold off on adding additional keywords for now.

**Pull request merged**